pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java

```diff
@@ -139,112 +153,293 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
+   * Updates a segment mapping if needed. In "downtime" mode.
+   * the segment mapping is set to the target mapping directly.
+   *
+   * In no-downtime mode, we check for couple of things and
+   * keep a certain number of serving replicas for the segment alive
+   * as specified in the rebalance configuration
+   *
+   * (1) if the number of common hosts between current and target
+   * mapping are enough to satisfy more than or equal to the number
+   * of serving replicas we should keep alive then we set to the
+   * target mapping directly (at one go) since this honors the no-downtime
+   * requirement
+   *
+   * (2) however, if there are not enough common hosts then we don't
+   * change the mapping directly as that will result in downtime. The
+   * mapping is updated incrementally (only one change for a given
+   * invocation of this method per segment) by removing a host
+   * from the current mapping and adding a host to it from
+   * target mapping.
+   *
+   * When the caller of this method returns (after going over all
+   * segments once), it persists the new ideal state and then determines
+   * if we have reached the target. If not, the process continues and we come
+   * here again for each method and check for steps (1) or (2) as applicable
+   *
+   * @param segmentId segment id
+   * @param currentIdealStateSegmentHosts map of this segment's hosts (replicas)
+   *                                     in current ideal state
+   * @param targetIdealStateSegmentHosts map of this segment's hosts (replicas)
+   *                                     in target ideal state
+   * @param idealStateToUpdate the ideal state that is updated as (caller
+   *                           passes this as a copy of current ideal state)
+   * @param rebalanceUserConfig rebalance configuration
    */
   @VisibleForTesting
-  public void updateSegmentIfNeeded(String segmentId, Map<String, String> srcMap, Map<String, String> targetMap,
-      IdealState idealState, Configuration rebalanceUserConfig) {
+  public void updateSegmentIfNeeded(String segmentId, Map<String, String> currentIdealStateSegmentHosts,
+      Map<String, String> targetIdealStateSegmentHosts, IdealState idealStateToUpdate,
+      Configuration rebalanceUserConfig) {
 
-    if (srcMap == null) {
+    LOGGER.debug("Will update instance map of segment: {}", segmentId);
+
+    if (currentIdealStateSegmentHosts == null) {
       //segment can be missing if retention manager has deleted it
-      LOGGER.info("Segment " + segmentId + " missing from current idealState. Skipping it.");
+      LOGGER.debug("Segment " + segmentId + " missing from current idealState. Skipping it.");
       return;
     }
 
-    if (rebalanceUserConfig
-        .getBoolean(RebalanceUserConfigConstants.DOWNTIME, RebalanceUserConfigConstants.DEFAULT_DOWNTIME)) {
-      setTargetState(idealState, segmentId, targetMap);
+    // dump additional detailed info for debugging
+    if (LOGGER.isDebugEnabled()) {
```

Comment:
do we really need this check?

Reply:
I thought we can check upfront to avoid running unnecessary logic that is needed to pretty print map by using joiner.